bin/publish_pod

```diff
@@ -1,10 +1,29 @@
 #!/bin/bash -eu
 
+# Usage: publish_pod [OPTIONS] PODSPEC_PATH
+#
+# OPTIONS:
+#  `--patch-cocoapods`:
+#     Apply a patch to work around issues with older deployment targets — see https://github.com/CocoaPods/CocoaPods/issues/12033
+#  `--allow-warnings`, `--synchronous`:
+#     Those options are passed to `pod trunk push` verbatim.
+#
+# Note: Use `--synchronous` if you have co-dependant podspecs in your repo and need to publish multiple pods at the same time.
```

Comment:
I wonder if it wouldn't make sense to automatically add `--synchronous` when multiple pod specs are used 🤔 

I can imagine a hypothetical scenario of pod specs that aren't dependent being pushed together, but I guess `--synchronous` wouldn't hurt either on that case?

Reply:
It would hurt the build time, because `--synchronous` implies that CocoaPods will clone the master Spec repo (which is **_huge_** —like multiple GBs—and quite long to clone) so that when it validates the second pod it can see the first and newly published pod immediately in the master spec repo instead of checking the CDN.

While without `--synchronous` it only checks the CDN (aka `trunk`) and doesn't need to clone (or restore cache of) the huge master spec repo.

---

One could argue that we don't have this level of refinement in our other helpers—for example our `bin/install_cocoapods` command unconditionally calls `restore_cache "$BUILDKITE_PIPELINE_SLUG-specs-repos"` and thus downloads the cache'd version of the CocoaPods master spec repo… even if in 95% of cases we don't need it at all (The `Podfile` of most repos should have removed the legacy `source 'https://github.com/CocoaPods/Specs.git'` line and migrated to use the CDN by now—but even if it has been years and years since CDN was introduced… very old app projects which were created even before that might never have taken the time to remove that legacy line though).

So in the same vein, maybe we should also modernize our `bin/install_cocoapods` to **only** call `restore_cache "$BUILDKITE_PIPELINE_SLUG-specs-repos"` if the `SPEC REPOS` section of the `Podfile.lock` contains the master spec repo URL (`https://github.com/CocoaPods/Specs.git`) explicitly (and then maybe emit a `buildkite-agent annotate --style warning` to suggest to migrate to CDN in those cases? 🤔 ), and skip the `restore_cache` if not? But that's an optimization for another PR, touching a different `bin/` helper than the ones at hand for solving the co-dependent pods validation issue.

Reply:
> It would hurt the build time

Do we have a case of non dependent pod specs being pushed together, though?
One could also argue that, if there's such need, you should call `publish_pod` separately anyway, while calling it for multiple pod specs would always imply on having them pushed with `--synchronous` 🤔 

> So in the same vein, maybe we should also modernize our bin/install_cocoapods

That sounds like a good thing to do anyway 👍

Reply:
Except that one might want to push one pod with `--allow-warnings` and the other without `--allow-warnings`, so in such case they wouldn't be able to call `publish_pod A.podspec B.podspec` as a single call.

That being said, I guess we could still consider making `publish_pod` automatically add `--synchronous` if `$# > 1` after consuming known flags… to at least make this common case automatic, all while still allowing to also provide it explicitly for cases where we'd want to call `publish_pod` on each pod separately (eg with different option for each).

---

The only drawback with that approach is that there wouldn't be an easy way to disable `--synchronous` if we wanted to publish multiple non-dépendant pods via a single call. In contrast, for the case of `--include-podspecs` (on `validate_podspec`) which is added implicitly if not provided, at least we can pass `--include-podspecs=''` to disable it, while there's no equivalent possibility if we were to make `--synchronous` implicit 🤔

Reply:
> to at least make this common case automatic

Yeah, that was my main point -- I see it more as the rule rather than the exception.

> Except that one might want to push one pod with --allow-warnings and the other without --allow-warnings, so in such case they wouldn't be able to call publish_pod A.podspec B.podspec as a single call.

True. It's hard to anticipate all these uses while being somewhat opinionated on how the commands should be used.
Perhaps it's best then to keep the options that can easily cover most cases and not block potential uses 🤷‍♂️